Fix b/c issue with folders for files and images of com_media from the 3.9.25 release

richard67 opened the pull request:
Pull Request for Issue #32567 .

### Summary of Changes

This pull request fixes the b/c issue with folders for files and images of com_media starting with something else than a letter a-z (upper or lower case) introduced with a security fix into the 3.9.25 release without introducing back the original security issue.

This is done by removing the check for the 1st character which was introduced to disable both full path disclosure by using a path like `./` or `.\` and break out of the parent folder by using `../` or `..\` (depending on your server's OS).

The usage of `./` or `.\` is handled instead by adding `'.'` to the exclude list for the parent folders:
https://github.com/joomla/joomla-cms/blob/754850e72495ca8cbdbc29c0e0debd51db26fce5/libraries/src/Form/Rule/FilePathRule.php#L55-L56

The breakout from the parent folder by using something like `../` or `..\` (depending on your OS) is already handled by the `Path::check($value)` a few lines further down. I've added the comment in line 69 to make that more clear:
https://github.com/joomla/joomla-cms/blob/754850e72495ca8cbdbc29c0e0debd51db26fce5/libraries/src/Form/Rule/FilePathRule.php#L69-L78

In addition to this fix, this PR also fixes the splitting of the value into parts for the exclude list check.

When on a Windows server. you perfectly can set and later use a path to images or files separates with either a Linux or a Windows directory separator.

In case if your site has been migrated from a Windows server e.g. for local development to the final Linux server, you still might have a `\` in a path, which would lead to false results when splitting the path for the exclude list check.

Therefore the change from `$path = explode('/', $value);` to `$path = preg_split('/[\/\\\\]/', $value);`.

### Testing Instructions

1. On a current staging or latest 3.9.x nightly build or 3.9.25, go to the media manager options in backend.
2. Either in the "Path to Files Folder" or for "Path to Images Folder" field, enter a path that starts with something else than a letter a to Z or A to Z.
3. Use the "Save" button.
Result: You **_can't_** save the changes, you get a red error message shown for the particular field.
4. Repeat steps 2 and 3 so you cover at least the following cases for both of these fields:
- Hidden folder like eg `.mysecretimages` as root.
- Root folder starting with a number, eg `1mississippi`, `2mississippi`, ...
- Root folder starting with a Unicode character, e.g. Cyrillic `Изображений`, Greece `Εικόνες`, ...
- Root folder starting with a symbol which is allowed for folder names on your server's OS, e.g. `-dada`, `_gaga`, ...

Result: Same as for step 3.

5. Apply the patch of this PR.
6. Repeat step 4, i.e. steps 2 and 3 at least for the cases mentioned in step 4.
Result: You **_can_** save the changes.
7. Enter a path starting with `./` or `.\` and use the "Save" button.
Result: You **_can't_** save the changes, you get a red error message shown for the particular field.
8. Enter a path which starts with `../` or contains `/../` (replace `/` by the actual directory separator of your server's OS) and use the "Save" button.
Result: You **_can't_** save the changes, you get a red error message shown for the particular field.
9. Enter a path which has one of the folders from the exclude list (see below) as the root folder, followed by a slash `/`, e.g. `administrator/`, and use the "Save" button.
Result: You **_can't_** save the changes, you get a red error message shown for the particular field.
10. Enter a path which has one of the folders from the exclude list (see below) as the root folder, followed by a backslash `\`, e.g. `administrator\`, and use the "Save" button.
Result: You **_can't_** save the changes, you get a red error message shown for the particular field.

The exclude list for field "Path to Files Folder" can be found here:
https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_media/config.xml#L41

The exclude list for field "Path to Images Folder" can be found here:
https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_media/config.xml#L52

### Actual result BEFORE applying this Pull Request

You **_can't_** save com_media settings when using a `Path to Files Folder` or a `Path to Images Folder` not starting with one of the letters a to z (case-insensitive).

This means eg the following is not allowed with 3.9.25 or current staging, which is a b/c break because it was allowed with 3.9.24 and older:
- Hidden folder like eg `.mysecretimages` as root.
- Root folder starting with a number.
- Root folder starting a Unicode character, e.g. Cyrillic, Greece, ...
- Root folder starting with a symbol which is allowed for folder names on your server's OS, e.g. `-` or `_`

### Expected result AFTER applying this Pull Request

You **_can_** save com_media settings when using a `Path to Files Folder` or a `Path to Images Folder` not starting with one of the letters a to z (case-insensitive).

You still can't save when one of these paths starts with something like `./` or `.\` on any OS or when it contains something like `../` on Linux or `..\` on Windows.

The check for the exclude list works with both kinds of directory separators `/`and `\`.

### Documentation Changes Required

None.

richard67 committed:
Fix FilePathRule b/c issue

PhilETaylor commented:
This is the perfect candidate for unit tests, being a validation rule, and being previously a security issue... Are there any unit tests that cover this ?

PhilETaylor commented on libraries/src/Form/Rule/FilePathRule.php:
```diff
@@ -49,15 +49,14 @@ public function test(\SimpleXMLElement $element, $value, $group = null, Registry
+		// Check the exclude setting
+		$path = preg_split('/[\/\\\\]/', $value);
 
 		if (!empty($exclude) && (in_array(strtolower($path[0]), $exclude) || empty($path[0])))
```
`$exclude` can never be empty now because you have added `.` to the array.

richard67 replied:
Right. Corrected.

richard67 committed:
Exclude list cannot be empty now

richard67 commented:
> Are there any unit tests that cover this ?

If there were any unit tests for this I would have extended them by new test cases, but I haven't found any, and creating completely new tests is beyond my knowledge (up to now).

PhilETaylor commented:
I will write the unit test. Please tell me, which of these do you believe should return false and which true:

```
./images
/images
.\images
../images
.../images
c:\images
\images
ftp://images
http://images
```

zero-24 commented:
>I will write the unit test. Please tell me, which of these do you believe should return false and which true:

Great, i did go througth the stuff and added a few more.

> ./images => false
> /images = true
> .\images  => false
> ../images  => false
> .../images  => false
> c:\images = true
> \images => false
> ftp://images  => false
> http://images  => false
> /media => false
> /administrator => false
> /4711images = true

PhilETaylor commented:
media and administrator are covered in the `won't fix` issue here https://github.com/joomla/joomla-cms/issues/32577

zero-24 commented:
Yes but that can be covered within the unit test too right?

richard67 commented:
> 
> 
> > I will write the unit test. Please tell me, which of these do you believe should return false and which true:
> 
> Great, i did go througth the stuff and added a few more.
> 
> > ./images => false
> > /images = true
> > .\images  => false
> > ../images  => false
> > .../images  => false
> > c:\images = true
> > \images => false
> > ftp://images  => false
> > http://images  => false
> > /media => false
> > /administrator => false
> > /4711images = true

@zero-24 Some of the results depend on how Path::clean works, and this is different on Linux or Windows. Have you tested the above on both kinds of OS, or is your list just based on your assumption how it should be?

I have tested it on Linux and i't wrong for some values.

PhilETaylor commented:
My unit test will be based on the current staging commit. 

You can then quickly merge it - because all the tests pass.

If the tests no longer pass once you have merged staging into your branch, then you have created a b/c break from staging (or fixed the previous b/c that 3.9.25 caused :-) )

Im pushing in 2 mins...

zero-24 commented:
> Have you tested the above on both kinds of OS, or is your list just based on your assumption how it should be?

based on my assumptions I might be wrong in some places.

richard67 commented:
> If the tests no longer pass once you have merged staging into your branch, then you have created a b/c break from staging (or fixed the previous b/c that 3.9.25 caused :-) )

They will break if you have a test for false being returned for the b/c that 3.9.25 caused?

PhilETaylor commented:
https://github.com/joomla/joomla-cms/pull/32723

richard67 commented:
@PhilETaylor You run them on Linux? Or Windows? Our testing environments run on Windows as far as I know, but the unit tests should also work on Linux. And there will be a differente due to current behaviour of `Path::check`.

PhilETaylor commented:
This is why they need to be merged, because then they will get run. 

I use a Mac.

PhilETaylor commented:
Also if the Joomla test suites are only being run on windows, then /facepalm ...

richard67 commented:
> This is why they need to be merged, because then they will get run.

They can be run from command line, too, and as I said, some of your tests might fail on one or the other OS.

richard67 commented:
@PhilETaylor Beside the above, what's the plan now? It needs to merge your PR before we can test and merge mine?

PhilETaylor commented:
> They can be run from command line, too,

They pass at the command line on my Mac, and it looks like the PR is passing unit testing too.

If the tests pass on one platform and fail on another, it would mean that the rule can return true when it should be false or false when it should be true... that's not a very good rule then. 

Thats not the tests fault. Thats the rules fault :-) 

> It needs to merge your PR before we can test and merge mine?

Thats not my decision. I just submit PR's and Joomla does what Joomla wants with them. The unit test submitted passes at the time it was created against staging so it would be easy for a maintainer to merge it once approved by 2 people - or merged without testing it ;-) 

I'll get a Joomla 4 one done too.

richard67 commented:
> If the tests pass on one platform and fail on another, it would mean that the rule can return true when it should be false or false when it should be true... that's not a very good rule then.
> 
> Thats not the tests fault. Thats the rules fault :-)

The rule used `Path::clean` before this PR and before the security fix, and it's that `Path::clean` which behaves different e.g. regarding the check for `..\` or `../`.

So if that is a mistake, it doesn't come from this PR here.

PhilETaylor commented:
Joomla 4 test here https://github.com/joomla/joomla-cms/pull/32724 which passes based on the current state of 4.0-dev and can be changed/improved at a later date

richard67 commented:
> Also if the Joomla test suites are only being run on windows, then /facepalm ...

@PhilETaylor I think I was wrong with that and they are running on Linux, at least some of them.

PhilETaylor commented:
AppVeyor is running windows. 

https://ci.appveyor.com/project/release-joomla/joomla-cms/builds/38276109/job/lpq1nwygs34vr7qk?fullLog=true

```
Build started
2SET PATH=C:\Program Files\OpenSSL;C:\tools\php;%PATH%
3SET COMPOSER_NO_INTERACTION=1
4SET PHP=1
5SET ANSICON=121x90 (121x90)
6git clone -q https://github.com/joomla/joomla-cms.git C:\projects\joomla-cms
7git fetch -q origin +refs/pull/32723/merge:
8git checkout -qf FETCH_HEAD
9Running
.....

```

richard67 commented:
> AppVeyor is running windows.

Yes, I got it now. Thanks.

PhilETaylor commented:
So to be clear. Is this still a b/c break? Are there any paths that used to work that will no longer work after this?

richard67 commented:
> So to be clear. Is this still a b/c break? Are there any paths that used to work that will no longer work after this?

No, this here should not be a b/c break so that it will make things not work which work without this PR, and if it is because I've mixed something, then I have to fix it.

But it will fix the b/c break from 3.9.25.

PhilETaylor commented:
applying this PR, against the unit tests (which are configured for staging, eg post 3.9.25 release) they fail:

```
PHPUnit 4.8.36 by Sebastian Bergmann and contributors.
Warning:	The Xdebug extension is not loaded
		No code coverage will be generated.

.F...........FFFFFF..FF

Time: 78 ms, Memory: 8.00MB

There were 9 failures:

1) FilePathRuleTest::testRule with data set #1 (false, SimpleXMLElement Object (...), '.images')
Failed asserting that true matches expected false.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

2) FilePathRuleTest::testRule with data set #13 (false, SimpleXMLElement Object (...), '4711images')
Failed asserting that true matches expected false.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

3) FilePathRuleTest::testRule with data set #14 (false, SimpleXMLElement Object (...), '1')
Failed asserting that true matches expected false.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

4) FilePathRuleTest::testRule with data set #15 (false, SimpleXMLElement Object (...), '_')
Failed asserting that true matches expected false.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

5) FilePathRuleTest::testRule with data set #16 (false, SimpleXMLElement Object (...), '*')
Failed asserting that true matches expected false.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

6) FilePathRuleTest::testRule with data set #17 (false, SimpleXMLElement Object (...), '%')
Failed asserting that true matches expected false.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

7) FilePathRuleTest::testRule with data set #18 (false, SimpleXMLElement Object (...), '://foo')
Failed asserting that true matches expected false.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

8) FilePathRuleTest::testRule with data set #21 (false, SimpleXMLElement Object (...), 'Εικόνες')
Failed asserting that true matches expected false.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

9) FilePathRuleTest::testRule with data set #22 (false, SimpleXMLElement Object (...), 'Изображений')
Failed asserting that true matches expected false.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

FAILURES!
Tests: 23, Assertions: 23, Failures: 9.
```

richard67 commented:
> applying this PR, against the unit tests (which are configured for staging, eg post 3.9.25 release) they fail:

@PhilETaylor All these are the fixes for the b/c introduced with 3.9.25, or not? What do your unit tests do on a 3.9.24?

PhilETaylor commented:
And this people.... is why Unit tests ARE SO IMPORTANT!

richard67 commented:
> And this people.... is why Unit tests ARE SO IMPORTANT!

@PhilETaylor I've never said the opposite. But as you can see I am not an expert in test (yet).

As said, if you run your tests as they are on a 3.9.24 they should fail in the same way like now when you run them on my PR, right?

PhilETaylor commented:
In order for Joomla 3.9.24 to pass the unit tests they need to be configured like this:

```
			array(true, $xml, ''),
			array(true, $xml, '.images'),
			array(false, $xml, './images'),
			array(false, $xml, '.\images'),
			array(false, $xml, '../images'),
			array(false, $xml, '.../images'),
			array(true, $xml, 'c:\images'),
			array(false, $xml, '\\images'), // Means \images
			array(true, $xml, 'ftp://images'),
			array(true, $xml, 'http://images'),
			array(false, $xml, '/media'),
			array(false, $xml, '/administrator'),
			array(false, $xml, '/4711images'),
			array(true, $xml, '4711images'),
			array(true, $xml, '1'),
			array(true, $xml, '_'),
			array(true, $xml, '*'),
			array(true, $xml, '%'),
			array(true, $xml, '://foo'),
			array(false, $xml, '/4711i/images'),
			array(false, $xml, '../4711i/images'),
			array(true, $xml, 'Εικόνες'),
			array(true, $xml, 'Изображений'),
```

So... what you need to do is tell me which of these, because of SECURITY REASONS are no longer allowed, and because of SECURITY REASONS you are choosing to change these from true/false to true/false therefore creating a b/c break. 

Because the above is the rules that pass on Joomla 3.9.24, they SHOULD pass on 3.9.25 and 3.9.26 else there has been a b.c break introduced.

PhilETaylor commented:
In order for Joomla 3.9.24 to pass the unit tests they need to be configured like this:

These all pass like this in Joomla 3.9.24
```
		    array(true, $xml, ''),
			array(true, $xml, '.images'),
			array(true, $xml, './images'),
			array(true, $xml, '.\images'),
			array(false, $xml, '../images'),
			array(false, $xml, '.../images'),
			array(true, $xml, 'c:\images'),
			array(true, $xml, '\\images'), // Means \images
			array(true, $xml, 'ftp://images'),
			array(true, $xml, 'http://images'),
			array(true, $xml, '/media'),
			array(true, $xml, '/administrator'),
			array(true, $xml, '/4711images'),
			array(true, $xml, '4711images'),
			array(true, $xml, '1'),
			array(true, $xml, '_'),
			array(true, $xml, '*'),
			array(true, $xml, '%'),
			array(true, $xml, '://foo'),
			array(true, $xml, '/4711i/images'),
			array(false, $xml, '../4711i/images'),
			array(true, $xml, 'Εικόνες'),
			array(true, $xml, 'Изображений'),
```

Then when I apply your PR to Joomla 3.9.24 I get failures:

```
There were 7 failures:

1) FilePathRuleTest::testRule with data set #2 (true, SimpleXMLElement Object (...), './images')
Failed asserting that false matches expected true.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

2) FilePathRuleTest::testRule with data set #3 (true, SimpleXMLElement Object (...), '.\images')
Failed asserting that false matches expected true.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

3) FilePathRuleTest::testRule with data set #7 (true, SimpleXMLElement Object (...), '\images')
Failed asserting that false matches expected true.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

4) FilePathRuleTest::testRule with data set #10 (true, SimpleXMLElement Object (...), '/media')
Failed asserting that false matches expected true.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

5) FilePathRuleTest::testRule with data set #11 (true, SimpleXMLElement Object (...), '/administrator')
Failed asserting that false matches expected true.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

6) FilePathRuleTest::testRule with data set #12 (true, SimpleXMLElement Object (...), '/4711images')
Failed asserting that false matches expected true.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

7) FilePathRuleTest::testRule with data set #19 (true, SimpleXMLElement Object (...), '/4711i/images')
Failed asserting that false matches expected true.

/Users/phil/Sites/JOOMLA/joomla3/tests/unit/suites/libraries/cms/form/rule/FilePathRuleTest.php:83

FAILURES!
Tests: 23, Assertions: 23, Failures: 7.

```

This tells me that your PR outputs something different than what Joomla 3.9.24 would have... 


So... what you need to do is tell me which of these, because of SECURITY REASONS are no longer allowed, and because of SECURITY REASONS you are choosing to change these from true/false to true/false therefore creating a b/c break. 

Because the above is the rules that pass on Joomla 3.9.24, they SHOULD pass on 3.9.25 and 3.9.26 else there has been a b.c break introduced somewhere.

PhilETaylor commented:
So we know /media and /administator are a b/c break - there is another issue about that with a pending `won't fix` on it... 

So the others that break are:

./images
.\images
\images
/4711images
/4711i/images

Which Im assuming is you making the security fix.... so it looks like the PR is acceptable, and the tests, once merged can be updated to take that into account.

richard67 commented:
> So the others that break are:
> 
> ./images
> .\images
> \images
> /4711images
> /4711i/images
> 
> Which Im assuming is you making the security fix.... **so it looks like the PR is acceptable**, and the tests, once merged can be updated to take that into account.

@PhilETaylor Yes, these are the intended ones by the security fix.

Thanks for your help and analysis.

I agree with you that it makes sense to merge the tests first and then update this one here so one can see by the changes in the tests in this PR what this PR does.

PhilETaylor commented:
I have tested this item :white_check_mark: successfully on a753e1d1022e534bc4fbd2ed5677ee1466bd34c9<br /><br />As per my comments above. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32718">issues.joomla.org/tracker/joomla-cms/32718</a>.</sub>

PhilETaylor commented:
> I agree with you that it makes sense to merge the tests first and then update this one here so one can see by the changes in the tests in this PR what this PR does.

remembering that the tests created today by me, are valid against the CURRENT staging, and not a released version (and so although they "pass" they "pass" validating the code added in 3.9.26 :-)... 

So yes. 1) Merge the working unit tests 2) merge that into this pr 3) update this PR by toggling the true/false on the tests to make them pass again and push this PR again ...

rdeutz commented:
We need to add changes the test for this PR, I would suggest for testing that testers apply first only the changes to the test and see if tests are failing.

**Instructions for testing**
* get the latest staging
* apply the changes to "tests/unit/suites/libraries/joomla/form/rule/FilePathRuleTest.php"
* execute the test -> should fail
* apply the complete PR
* execute the test -> should not fail

How the execute the tests?
* go into the root of the checked out repository
* do a "composer install"
* do a "phpunit --filter FilePathRuleTest"

**As of the writing of this comment changes for the test are not within the PR**

richard67 commented:
In some 2 hours I will have some time to update this PR by the new tests and adjust the tests. Please wait with the above testing until l’ve done that.

richard67 committed:
Merge branch 'staging' into staging-fix-files-and-images-folder-bc-issue

richard67 committed:
Adapt new unit tests from PR 32723 to the changes of this one here

richard67 commented:
@PhilETaylor I've updated this PR and adjusted the tests.

But I've noticed a mistake in your test cases: They are testing `/media` and `/administrator`, but these don't fail due to the excluded core folder list, these fail because they start with a directory separator, i.e. absolute path, i.e. `$path[0]` is an empty string after the split of `$value`. So I'll add tests for `media` and `administrator` (without ´/` at the beginning). You should do the same in your PR #32724 .

richard67 committed:
Add test cases for excluded core folders

richard67 committed:
Add exclude parameter to the form field for unit tests

richard67 commented:
Tests failing now for unrelated reasons. Will restart from time to time until ok.

richard67 commented:
Ready for testing.

@PhilETaylor Could you test again? Thanks in advance, and again thanks for good team work.

richard67 commented:
I've restored @PhilETaylor 's test result because all changes after that test were only for the unit tests, not for the rule code.

PhilETaylor commented:
All your work looks ok, my only note is that the Joomla 3 test suite's PHPUnit version is incompatible with PHP 8 and so this has not been tested (and actually fails in drone) in PHP 8.

```
Fatal error: Uncaught Error: Call to undefined function each() in /drone/src/libraries/vendor/phpunit/phpunit/src/Util/Getopt.php:38
```

That is code for "This PHPUnit version is incompatible with PHP 8" ;-)

richard67 commented:
> 
> All your work looks ok, my only note is that the Joomla 3 test suite's PHPUnit version is incompatible with PHP 8 and so this has not been tested (and actually fails in drone) in PHP 8.

@PhilETaylor But I get it right that this is not related to my PR but also happens on current staging?

PhilETaylor commented:
Correct. This is not your fault, I was nearly translating that error message into English so that you and others dont waste hours trying to fix the PHP 8 unit testing for Joomla 3. It will never work without upgrading PHPUnit and that then will break all the tests as the current PHPUnit version that is compatible with PHP8 is not compatible with the tests (and older PHP versions).

So, just ignore it. 

Hint: I wasted hours trying to "fix" the unfixable PHPUnit error message :)

alikon commented:
I have tested this item :white_check_mark: successfully on 4566d0e44f47e1ae92e332f3fd6f8cd9d783d342<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32718">issues.joomla.org/tracker/joomla-cms/32718</a>.</sub>

alikon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32718">issues.joomla.org/tracker/joomla-cms/32718</a>.</sub>

PhilETaylor commented:
Thank you @richard67 @alikon

richard67 commented:
@PhilETaylor Thank you Phil.

BMan54 commented:
Be glad to volunteer and help test.  My desktop is Windows 10 Pro. I have Hyper V installed with Centos 7, but I'll gladly install any flavor of Linux - I've built LAMP stacks on both Debian and RH derivatives.

PhilETaylor commented:
I just feel foul of this a-z issue in Joomla 4 - does anyone know where @wilsonge is with merging up from staging the commits? Is it documented anywhere? is it still a one man effort?

zero-24 commented:
> I just fell foul of this a-z issue in Joomla 4 - does anyone know where @wilsonge is with merging up from staging the commits?

hmm merging up usually happens after a release. So staging -> 3.10-dev -> 4.0-dev and it seems George [merged the latest stuff up 7 days from today](https://github.com/joomla/joomla-cms/commit/321def11e90b7e9e2ffc9d5e6744c06a58b4a3b5).
Following that logic would mean we would wait for 3.9.26 to be released, merged up to 3.10-dev and than merged up to 4.0-dev.

>  Is it documented anywhere?

What kind of docs are you looking for.

> is it still a one man effort?

Merging up stuff is usually done by the release leads so all possible merge conflicts can be reviewed by that person too but sometimes other maintainers take that over too.

PhilETaylor commented:
Well my concern is things might get missed.

However it seems the commits that I was concerned about were merged after the 3.9.25 was released, so from what you say they should not yet be in Joomla 4.

zero-24 commented:
> However it seems the commits that I was concerned about were merged after the 3.9.25 was released, so from what you say they should not yet be in Joomla 4.

Now I'm confused what commits are you concerned about than? When its about this PR here it will be shipped tomorrow with 3.9.26 and has not been merged up yet?

PhilETaylor commented:
Long story. Im having a bad day :)
